src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -36,43 +33,38 @@ internal sealed class FileChangeWatcher
         /// For performance and correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block and add any work to
         /// the end of the queue.
         /// </summary>
-        private Task<IVsAsyncFileChangeEx> _taskQueue;
-        private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
-            async (precedingTask, state) =>
-            {
-                var action = (Func<IVsAsyncFileChangeEx, Task>)state;
-                await action(precedingTask.Result).ConfigureAwait(false);
-                return precedingTask.Result;
-            };
-
-        public FileChangeWatcher(Task<IVsAsyncFileChangeEx> fileChangeService)
-            => _taskQueue = fileChangeService;
+        private readonly AsyncBatchingWorkQueue<WatcherOperation> _taskQueue;
 
-        private void EnqueueWork(Func<IVsAsyncFileChangeEx, Task> action)
+        public FileChangeWatcher(
+            IAsynchronousOperationListenerProvider listenerProvider,
+            Task<IVsAsyncFileChangeEx> fileChangeService)
         {
-            lock (_taskQueueGate)
-            {
-                _taskQueue = _taskQueue.ContinueWith(
-                    _executeActionDelegate,
-                    action,
-                    CancellationToken.None,
-                    TaskContinuationOptions.None,
-                    TaskScheduler.Default).Unwrap();
-            }
+            _fileChangeService = fileChangeService;
+            _taskQueue = new AsyncBatchingWorkQueue<WatcherOperation>(
+                TimeSpan.Zero,
+                ProcessBatchAsync,
+                listenerProvider.GetListener(FeatureAttribute.Workspace),
+                CancellationToken.None);
```

Comment:
consider taking in a ThreadingContext and using the `.DisposalToken` here.

Reply:
The current behavior intentionally matches the original behavior of this work queue. It might be safe to change, but I'm not comfortable making that change as part of the targeted fix here.

Reply:
sgtm.

Reply:
@sharwell @CyrusNajmabadi Want to file a bug for tracking to not lose this? This looks like a good suggestion, but I agree with the deferral.